gemini/responsible-ai/react_rag_attacks_mitigations_examples.ipynb

```diff
@@ -0,0 +1,1086 @@
+    "! pip install --user --quiet pdf2image"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0k92o64LTExx",
+   "metadata": {
+    "id": "0k92o64LTExx"
+   },
+   "source": [
+    "**The below code block is required to restart the runtime in colab after installing required dependencies.**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "vKIxInAvQt1d",
+   "metadata": {
+    "id": "vKIxInAvQt1d"
+   },
+   "outputs": [],
+   "source": [
+    "# # Automatically restart kernel after installs so that your environment can access the new packages\n",
+    "import IPython\n",
+    "\n",
+    "app = IPython.Application.instance()\n",
+    "app.kernel.do_shutdown(True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "DxpUqVVa3EgL",
+   "metadata": {
+    "id": "DxpUqVVa3EgL"
+   },
+   "source": [
+    "**Specify project and location to be used by this notebook and where to make the API calls. @Capstone team - replace with a project accessible to you with the required API services enabled**"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "L0OB-trIAGAy",
+   "metadata": {
+    "id": "L0OB-trIAGAy"
+   },
+   "source": [
+    "### Project and Authentication\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "jVBQF2w9iHHM",
+   "metadata": {
+    "id": "jVBQF2w9iHHM"
+   },
+   "outputs": [],
+   "source": [
+    "# Provide your Google Cloud project and region\n",
+    "project_id = \"your-project-123456\"  # @param {type:\"string\"}\n",
+    "location = \"us-central1\"  # @param {type:\"string\"}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "MOja8y76ANRM",
+   "metadata": {
+    "id": "MOja8y76ANRM"
+   },
+   "outputs": [],
+   "source": [
+    "# Authenticate the user if you use Colab\n",
+    "from google.colab import auth as google_auth\n",
+    "\n",
+    "google_auth.authenticate_user()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ZwWKigcJUmS_",
+   "metadata": {
+    "id": "ZwWKigcJUmS_"
+   },
+   "source": [
+    "## ReAct - Agent Security\n",
+    "\n",
+    "**Reasoning and Acting** (ReAct)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "FQxrMwVvsg-q",
+   "metadata": {
+    "id": "FQxrMwVvsg-q"
+   },
+   "source": [
+    "![mitigations-diagram.png](https://storage.googleapis.com/github-repo/responsible-ai/intro_genai_security/react.png)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "TjqiFJJe6Q5M",
+   "metadata": {
+    "id": "TjqiFJJe6Q5M"
+   },
+   "source": [
+    "### Agent Tools\n",
+    "Defining the tools use by the agent as simple python function. In real life this can be API calls"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "_yyBozpv6oQp",
+   "metadata": {
+    "id": "_yyBozpv6oQp"
+   },
+   "outputs": [],
+   "source": [
+    "import random\n",
+    "\n",
+    "\n",
+    "def weather_city(city):\n",
```

Comment:
Add typing for parameters and return values

Reply:
done